channeldb/invoice_preimages.go

```diff
@@ -0,0 +1,99 @@
+package channeldb
+
+import (
+  "bytes"
+  "crypto/sha256"
+  "fmt"
+
+  "github.com/btcsuite/btcd/chaincfg/chainhash"
+  "github.com/lightningnetwork/lnd/extpreimage"
+)
+
+// InvoiceRegistry is a registry for storing invoices. It is a
+// simplified interface of htlcswitch.InvoiceDatabase for use
+// in this package.
+type InvoiceRegistry interface {
+  AddInvoicePreimage(chainhash.Hash, [32]byte) error
+}
+
+// GetPaymentHash retrieves the payment hash for a given invoice,
+// either by calculating it from the preimage, or using the given
+// hash for invoices with external preimages.
+func (c *ContractTerm) GetPaymentHash() ([32]byte, error) {
+  var zeroHash [32]byte
+  var paymentHash [32]byte
+  var zeroPreimage [32]byte
+
+  if c.ExternalPreimage {
+    if bytes.Equal(c.PaymentHash[:], zeroHash[:]) {
+      return zeroHash, fmt.Errorf("Invoices with ExternalPreimage must " +
+        "have a locally defined PaymentHash.")
+    }
+
+    // For external preimages, we rely on a provided hash
+    paymentHash = c.PaymentHash
+  } else {
+    if bytes.Equal(c.PaymentPreimage[:], zeroPreimage[:]) {
+      return zeroHash, fmt.Errorf("Invoices must have a preimage or" +
+        "use ExternalPreimages")
+    }
+
+    // For local preimages, we calculate the hash ourselves
+    paymentHash = sha256.Sum256(c.PaymentPreimage[:])
+  }
+
+  return paymentHash, nil
+}
+
+// GetPaymentPreimage retrieves the preimage for a given invoice,
+// either by pulling it directly from the invoice, or by retrieving
+// it from the external preimage service if it is an external preimage
+// invoice.
+func (c *ContractTerm) GetPaymentPreimage(timeLock uint32, currentHeight uint32,
```

Comment:
Is there a way to document the return values to that we can say (preimage, temp error, perm error)